File: core/src/main/java/cz/dynawest/jtexy/util/ChildrenHTMLWriter.java

package cz.dynawest.jtexy.util;

import java.io.IOException;
import java.io.Writer;
import org.dom4j.io.HTMLWriter;
import org.dom4j.io.OutputFormat;
import org.xml.sax.Attributes;
import org.xml.sax.SAXException;

/**
 * This mutation of HTML writer will only write child nodes of the root.
 * E.g.:   <p>Hi</p> instead of <!DOCTYPE ...><html><p>Hi</p></html>
 *
 * @author Ondrej Zizka
 */
final class ChildrenHTMLWriter extends HTMLWriter {

	int depth = 0;

	Writer activeWriter = voidWriter;
	
	Writer realWriter;
	
	// No-op writer for nodes which will not be written.
	private static Writer voidWriter = new Writer() {
		@Override	public void write(char[] cbuf, int off, int len) throws IOException {	}
		@Override	public void flush() throws IOException { }
		@Override	public void close() throws IOException { }
	};


	// -- Const --

	public ChildrenHTMLWriter(Writer writer, OutputFormat format) {
		super(voidWriter, format);
		this.realWriter = writer;
	}

	public ChildrenHTMLWriter(Writer writer) {
		super(voidWriter);
		this.realWriter = writer;
	}


	// -- Overrides --

	/** Upon root element start, switch to real writer. */
	@Override
	public void startElement(String namespaceURI, String localName, String qName, Attributes attributes) throws SAXException {
		super.startElement(namespaceURI, localName, qName, attributes);
		if( ++depth == 1 )
			activeWriter = realWriter;
	}

	/** Upon root element end, switch back to void writer. */
	@Override
	public void endElement(String namespaceURI, String localName, String qName) throws SAXException {
		if( --depth == 0 )
			activeWriter = voidWriter;
		super.endElement(namespaceURI, localName, qName);
	}









}

File: core/src/main/java/cz/dynawest/jtexy/util/ChildrenHTMLWriter.kt
package cz.dynawest.jtexy.util

import org.dom4j.io.HTMLWriter
import org.dom4j.io.OutputFormat
import org.xml.sax.Attributes
import org.xml.sax.SAXException
import java.io.*

/**
 * This mutation of HTML writer will only write child nodes of the root.
 * E.g.:
 *
 *Hi instead of  <html>
 *
 *Hi</html>
 *
 * @author Ondrej Zizka
 */
internal class ChildrenHTMLWriter : HTMLWriter {
    var depth = 0
    var activeWriter = voidWriter
    var realWriter: Writer

    // -- Const --
    constructor(writer: Writer, format: OutputFormat?) : super(voidWriter, format) {
        realWriter = writer
    }

    constructor(writer: Writer) : super(voidWriter) {
        realWriter = writer
    }
    // -- Overrides --
    /** Upon root element start, switch to real writer.  */
    @Throws(SAXException::class)
    override fun startElement(namespaceURI: String, localName: String, qName: String, attributes: Attributes) {
        super.startElement(namespaceURI, localName, qName, attributes)
        if (++depth == 1) activeWriter = realWriter
    }

    /** Upon root element end, switch back to void writer.  */
    @Throws(SAXException::class)
    override fun endElement(namespaceURI: String, localName: String, qName: String) {
        if (--depth == 0) activeWriter = voidWriter
        super.endElement(namespaceURI, localName, qName)
    }

    companion object {
        // No-op writer for nodes which will not be written.
        private val voidWriter: Writer = object : Writer() {
            @Throws(IOException::class)
            override fun write(cbuf: CharArray, off: Int, len: Int) {
            }

            @Throws(IOException::class)
            override fun flush() {
            }

            @Throws(IOException::class)
            override fun close() {
            }
        }
    }
}
